workspaces/adventure-pack/goodies/typescript/Iterator.prototype.flatMap/index.test.ts

```diff
@@ -0,0 +1,195 @@
+      -1, -2, -5, -10, 5, 10, 3, 6, 7, 14,
+    ]);
+  });
+
+  it("can flatten a String's characters", () => {
+    const s = "hello";
+    const callback = function* (c: string): Generator<string, void, void> {
+      yield c;
+      yield c.toUpperCase();
+    };
+
+    const flatMapResult = s[Symbol.iterator]().flatMap(callback);
+
+    expect([...flatMapResult]).toStrictEqual(Array.from("hHeElLlLoO"));
+  });
+
+  it("can flatten a Map's entries()", () => {
+    const map = new Map([
+      ["apple", "green"],
+      ["banana", "yellow"],
+      ["grape", "purple"],
+    ]);
+    const generator = function* ([fruit, color]: [string, string]): Generator<
+      string,
+      void,
+      void
+    > {
+      yield `I found a ${color} ${fruit}.`;
+      yield `I love ${color} ${fruit}s, *BITE*.`;
+    };
+
+    const flatMapResult = map.entries().flatMap(generator);
+
+    expect([...flatMapResult]).toStrictEqual([
+      "I found a green apple.",
+      "I love green apples, *BITE*.",
+      "I found a yellow banana.",
+      "I love yellow bananas, *BITE*.",
+      "I found a purple grape.",
+      "I love purple grapes, *BITE*.",
+    ]);
+  });
+
+  it("can flatten a Set's values()", () => {
+    const set = new Set([2, 2, -4, -4, 7, 7, 9, 9, -10, -10]);
+    const callback = function* (
+      p: number,
+    ): Generator<number | boolean, void, void> {
+      yield p;
+      yield p % 2 === 0;
+    };
+
+    const flatMapResult = set.values().flatMap(callback);
+
+    expect([...flatMapResult]).toStrictEqual([
+      2,
+      true,
+      -4,
+      true,
+      7,
+      false,
+      9,
+      false,
+      -10,
+      true,
+    ]);
+  });
+
+  it("can flatten a Generator object", () => {
+    const factory = function* (): Generator<string, void, void> {
+      yield "o";
+      yield "t";
+      yield "t";
+      yield "e";
+      yield "r";
+    };
+
+    const callback = function* (c: string): Generator<string, void, void> {
+      yield c;
+      yield "!";
+    };
+
+    const flatMapResult = factory().flatMap(callback);
+
+    expect([...flatMapResult]).toStrictEqual(Array.from("o!t!t!e!r!"));
+  });
+
+  it("callback yields elements with their respective indices", () => {
+    const iterator = ["🍞", "🥓", "🥬", "🍅"].values();
+    const callback = function* (
+      c: string,
+      index: number,
+    ): Generator<string, void, void> {
+      yield `${index}: ${c}`;
+    };
+
+    const flatMapResult = iterator.flatMap(callback);
+    expect([...flatMapResult]).toStrictEqual([
+      "0: 🍞",
+      "1: 🥓",
+      "2: 🥬",
+      "3: 🍅",
+    ]);
+  });
+
+  it.each([
+    [].values(),
+    new Set().values(),
+    new Map().values(),
+    (function* () {})(),
+  ])(
+    "returns an empty iterator when called on an empty iterator",
+    <T>(iterator: IterableIterator<T>) => {
+      expect(
+        iterator
+          .flatMap(function* <T>(p: T): Generator<T, void, void> {
+            yield p;
```

Comment:
Fwiw, when we only yield once from a `.flatMap` callback it's not super interesting and it could have been a `.map` instead.